python/hail/methods/misc.py

```diff
@@ -195,61 +194,79 @@ def rename_duplicates(dataset, name='unique_id'):
     """Filter rows with an interval or list of intervals.
 
-    .. note::
-
-        Requires the dataset to have a single partition key of type
-        :class:`.tlocus`.
-
     Examples
     --------
 
     Filter to loci falling within one interval:
 
-    >>> ds_result = hl.filter_intervals(dataset, hl.Interval.parse('17:38449840-38530994'))
+    >>> ds_result = hl.filter_intervals(dataset, hl.parse_locus_interval('17:38449840-38530994'))
 
     Remove all loci within list of intervals:
 
-    >>> intervals = [hl.Interval.parse(x) for x in ['1:50M-75M', '2:START-400000', '3-22']]
+    >>> intervals = [hl.parse_locus_interval(x) for x in ['1:50M-75M', '2:START-400000', '3-22']]
     >>> ds_result = hl.filter_intervals(dataset, intervals)
 
     Notes
     -----
     This method takes an argument of :class:`.Interval` or list of
-    :class:`.Interval`.
-
-    Based on the ``keep`` argument, this method will either restrict to loci in
-    the supplied interval ranges, or remove all loci in those ranges.  Note that
-    intervals are left-inclusive, and right-exclusive; the below interval
-    includes the locus ``15:100000`` but not ``15:101000``.
+    :class:`.Interval` with a point type that is the same as the type of
+    the row partition key.
 
-    >>> interval = hl.Interval.parse('15:100000-101000')
+    Based on the ``keep`` argument, this method will either restrict to points
+    in the supplied interval ranges, or remove all rows in those ranges.
 
     When ``keep=True``, partitions that don't overlap any supplied interval
     will not be loaded at all.  This enables :func:`.filter_intervals` to be
-    used for reasonably low-latency queries of small ranges of the genome, even
+    used for reasonably low-latency queries of small ranges of the dataset, even
     on large datasets.
 
     Parameters
     ----------
+    ds : :class:`.MatrixTable`
+        Dataset.
     intervals : :class:`.Interval` or :obj:`list` of :class:`.Interval`
-        Interval(s) to filter on.
+        Interval(s) to filter on. If there is only one row partition key, the
+        point type of the interval can be the type of the first partition key.
+        Otherwise, the interval point type must be a :class:`.Struct` matching
+        the row partition key schema.
     keep : :obj:`bool`
-        If ``True``, keep only loci fall within any interval in `intervals`. If
-        ``False``, keep only loci that fall outside all intervals in
+        If ``True``, keep only rows that fall within any interval in `intervals`.
+        If ``False``, keep only rows that fall outside all intervals in
         `intervals`.
 
     Returns
     -------
     :class:`.MatrixTable`
     """
 
-    require_partition_key_locus(ds, 'filter_intervals')
-
-    intervals = wrap_to_list(intervals)
-    jmt = Env.hail().methods.FilterIntervals.apply(ds._jvds, [x._jrep for x in intervals], keep)
+    intervals = eval_expr(intervals)
```

Comment:
check the `intervals.dtype.element_type.point_type` here. Then the type error only needs to happen once, and up here.

Reply:
and set a boolean flag `needs_wrapper` that tells you whether to wrap it in a struct or not